File: backend/src/modules/openai-compatible-files/file-storage.ts
import { randomUUID } from 'node:crypto'
import { existsSync, mkdirSync } from 'node:fs'
import { unlink } from 'node:fs/promises'
import { dirname, relative, resolve } from 'node:path'

import { runtimeConfig } from '../../config/runtime.js'

export const openAICompatibleFileMaxBytes = 512 * 1024 * 1024
export const openAICompatibleBridgeFileMaxBytes = 32 * 1024 * 1024

export function newOpenAICompatibleFileId(): string {
  return `file-${Date.now().toString(36)}-${randomUUID().replace(/-/g, '').slice(0, 20)}`
}

export function storageKeyForOpenAICompatibleFile(fileId: string): string {
  const safeId = safeStorageSegment(fileId)
  const shard = safeId.slice(0, 8) || 'default'
  return `files/${shard}/${safeId}`
}

export function openAICompatibleFileObjectPath(storageKey: string): string {
  const root = openAICompatibleFilesRoot()
  const target = resolve(root, storageKey)
  const relativePath = relative(root, target)
  if (relativePath.startsWith('..') || relativePath === '' || /^[A-Za-z]:/.test(relativePath)) {
    throw new Error('OpenAI compatible file storage key escaped storage root')
  }
  return target
}

export function ensureOpenAICompatibleFileObjectParent(storageKey: string): string {
  const filePath = openAICompatibleFileObjectPath(storageKey)
  mkdirSync(dirname(filePath), { recursive: true })
  return filePath
}

export async function removeOpenAICompatibleFileObject(storageKey: string): Promise<void> {
  const filePath = openAICompatibleFileObjectPath(storageKey)
  if (!existsSync(filePath)) return
  await unlink(filePath)
}

export function mediaTypeFromFilename(filename: string | undefined): string | undefined {
  const lower = filename?.trim().toLowerCase()
  if (!lower) return undefined
  if (lower.endsWith('.pdf')) return 'application/pdf'
  if (lower.endsWith('.txt')) return 'text/plain'
  if (lower.endsWith('.md') || lower.endsWith('.markdown')) return 'text/markdown'
  if (lower.endsWith('.csv')) return 'text/csv'
  if (lower.endsWith('.json')) return 'application/json'
  if (lower.endsWith('.c')) return 'text/x-c'
  if (lower.endsWith('.cpp') || lower.endsWith('.cc') || lower.endsWith('.cxx')) return 'text/x-c++'
  if (lower.endsWith('.cs')) return 'text/x-csharp'
  if (lower.endsWith('.css')) return 'text/css'
  if (lower.endsWith('.go')) return 'text/x-golang'
  if (lower.endsWith('.html') || lower.endsWith('.htm')) return 'text/html'
  if (lower.endsWith('.java')) return 'text/x-java'
  if (lower.endsWith('.js') || lower.endsWith('.mjs') || lower.endsWith('.cjs')) return 'text/javascript'
  if (lower.endsWith('.php')) return 'text/x-php'
  if (lower.endsWith('.py')) return 'text/x-python'
  if (lower.endsWith('.rb')) return 'text/x-ruby'
  if (lower.endsWith('.tex')) return 'text/x-tex'
  if (lower.endsWith('.ts') || lower.endsWith('.tsx')) return 'application/typescript'
  if (lower.endsWith('.sh')) return 'application/x-sh'
  if (lower.endsWith('.png')) return 'image/png'
  if (lower.endsWith('.jpg') || lower.endsWith('.jpeg')) return 'image/jpeg'
  if (lower.endsWith('.gif')) return 'image/gif'
  if (lower.endsWith('.webp')) return 'image/webp'
  return undefined
}

export function normalizeOpenAICompatibleFileMediaType(value: string | undefined, filename: string | undefined): string | undefined {
  const mediaType = value?.split(';', 1)[0]?.trim().toLowerCase()
  if (mediaType && mediaType !== 'application/octet-stream') return mediaType
  return mediaTypeFromFilename(filename)
}

function openAICompatibleFilesRoot(): string {
  const root = resolve(runtimeConfig.openAICompatibleFilesRoot)
  mkdirSync(root, { recursive: true })
  return root
}

function safeStorageSegment(value: string): string {
  return value.replace(/[^A-Za-z0-9_.-]/g, '_').slice(0, 160) || 'file'
}

File: backend-go/projects/gateway/internal/openaicompat/filestorage.go
package openaicompat

import (
	"errors"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// MaxTextIndexBytes is re-exported next to the storage helpers for
// convenience of embedders (text-indexer keeps its own constant too).
const MaxTextIndexBytes = 2 * 1024 * 1024

// StorageKeyEscapeError mirrors the openAICompatibleFileObjectPath throw
// "OpenAI compatible file storage key escaped storage root".
var StorageKeyEscapeError = errors.New("OpenAI compatible file storage key escaped storage root")

var storageUnsafeCharacters = regexp.MustCompile(`[^A-Za-z0-9_.-]`)

// NewFileID returns a fresh file id using the default generator
// (file-<base36 ms>-<20 hex>).
func NewFileID(now time.Time) string { return newOpenAICompatibleFileID(now) }

// StorageKeyForFile mirrors storageKeyForOpenAICompatibleFile:
// files/<8-char shard>/<sanitized id>. The shard comes from the raw sanitized
// id (before its own "file" fallback), so an empty id lands in "default".
func StorageKeyForFile(fileID string) string {
	safeID := safeStorageSegment(fileID)
	shard := safeID
	if len([]rune(shard)) > 8 {
		shard = string([]rune(shard)[:8])
	}
	if shard == "" {
		shard = "default"
	}
	return "files/" + shard + "/" + safeID
}

// safeStorageSegment mirrors safeStorageSegment: replace anything outside
// [A-Za-z0-9_.-] with '_', clamp to 160 characters, fall back to "file".
func safeStorageSegment(value string) string {
	replaced := storageUnsafeCharacters.ReplaceAllString(value, "_")
	runes := []rune(replaced)
	if len(runes) > 160 {
		runes = runes[:160]
	}
	segment := string(runes)
	if segment == "" {
		return "file"
	}
	return segment
}

// isRootAnchoredKey reports keys that start at a filesystem root: absolute
// paths, volume-less rooted paths (backslash-x, forward-slash-x) or drive
// paths (C:x).
func isRootAnchoredKey(storageKey string) bool {
	if filepath.IsAbs(storageKey) || strings.HasPrefix(storageKey, "/") || strings.HasPrefix(storageKey, "\\") {
		return true
	}
	return len(storageKey) >= 2 && storageKey[1] == ':'
}

// FileObjectPath mirrors openAICompatibleFileObjectPath: resolve the storage
// key under root and refuse anything that escapes the root (Node checks the
// relative path for "..", an empty result or a Windows drive prefix).
func FileObjectPath(root, storageKey string) (string, error) {
	rootPath, err := filepath.Abs(root)
	if err != nil {
		return "", err
	}
	rootPath = filepath.Clean(rootPath)
	var target string
	if isRootAnchoredKey(storageKey) {
		// Node resolve(root, "/x") jumps to the drive root of root and the
		// subsequent relative check rejects it; mirror that rejection.
		target = filepath.Clean(storageKey)
	} else {
		target = filepath.Clean(filepath.Join(rootPath, storageKey))
	}
	relativePath, err := filepath.Rel(rootPath, target)
	if err != nil {
		return "", StorageKeyEscapeError
	}
	if relativePath == ".." || strings.HasPrefix(relativePath, ".."+string(filepath.Separator)) ||
		relativePath == "." || relativePath == "" || filepath.IsAbs(relativePath) {
		return "", StorageKeyEscapeError
	}
	return target, nil
}

// EnsureFileObjectParent mirrors ensureOpenAICompatibleFileObjectParent.
func EnsureFileObjectParent(root, storageKey string) (string, error) {
	filePath, err := FileObjectPath(root, storageKey)
	if err != nil {
		return "", err
	}
	if err := os.MkdirAll(filepath.Dir(filePath), 0o755); err != nil {
		return "", err
	}
	return filePath, nil
}

// RemoveFileObject mirrors removeOpenAICompatibleFileObject: missing files
// are not an error.
func RemoveFileObject(root, storageKey string) error {
	filePath, err := FileObjectPath(root, storageKey)
	if err != nil {
		return err
	}
	if _, err := os.Stat(filePath); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil
		}
		return err
	}
	return os.Remove(filePath)
}

// MediaTypeFromFilename mirrors mediaTypeFromFilename (exact suffix table and
// precedence order).
func MediaTypeFromFilename(filename string) string {
	lower := strings.ToLower(strings.TrimSpace(filename))
	if lower == "" {
		return ""
	}
	switch {
	case strings.HasSuffix(lower, ".pdf"):
		return "application/pdf"
	case strings.HasSuffix(lower, ".txt"):
		return "text/plain"
	case strings.HasSuffix(lower, ".md"), strings.HasSuffix(lower, ".markdown"):
		return "text/markdown"
	case strings.HasSuffix(lower, ".csv"):
		return "text/csv"
	case strings.HasSuffix(lower, ".json"):
		return "application/json"
	case strings.HasSuffix(lower, ".c"):
		return "text/x-c"
	case strings.HasSuffix(lower, ".cpp"), strings.HasSuffix(lower, ".cc"), strings.HasSuffix(lower, ".cxx"):
		return "text/x-c++"
	case strings.HasSuffix(lower, ".cs"):
		return "text/x-csharp"
	case strings.HasSuffix(lower, ".css"):
		return "text/css"
	case strings.HasSuffix(lower, ".go"):
		return "text/x-golang"
	case strings.HasSuffix(lower, ".html"), strings.HasSuffix(lower, ".htm"):
		return "text/html"
	case strings.HasSuffix(lower, ".java"):
		return "text/x-java"
	case strings.HasSuffix(lower, ".js"), strings.HasSuffix(lower, ".mjs"), strings.HasSuffix(lower, ".cjs"):
		return "text/javascript"
	case strings.HasSuffix(lower, ".php"):
		return "text/x-php"
	case strings.HasSuffix(lower, ".py"):
		return "text/x-python"
	case strings.HasSuffix(lower, ".rb"):
		return "text/x-ruby"
	case strings.HasSuffix(lower, ".tex"):
		return "text/x-tex"
	case strings.HasSuffix(lower, ".ts"), strings.HasSuffix(lower, ".tsx"):
		return "application/typescript"
	case strings.HasSuffix(lower, ".sh"):
		return "application/x-sh"
	case strings.HasSuffix(lower, ".png"):
		return "image/png"
	case strings.HasSuffix(lower, ".jpg"), strings.HasSuffix(lower, ".jpeg"):
		return "image/jpeg"
	case strings.HasSuffix(lower, ".gif"):
		return "image/gif"
	case strings.HasSuffix(lower, ".webp"):
		return "image/webp"
	default:
		return ""
	}
}

// NormalizeFileMediaType mirrors normalizeOpenAICompatibleFileMediaType.
func NormalizeFileMediaType(value, filename string) string {
	mediaType := strings.ToLower(strings.TrimSpace(strings.SplitN(value, ";", 2)[0]))
	if mediaType != "" && mediaType != "application/octet-stream" {
		return mediaType
	}
	return MediaTypeFromFilename(filename)
}

// NormalizedUploadFilename mirrors normalizedUploadFilename: last segment
// across / and \ separators, trimmed, defaulting to "upload".
func NormalizedUploadFilename(value string) string {
	segments := strings.Split(value, "/")
	last := segments[len(segments)-1]
	if index := strings.LastIndex(last, "\\"); index >= 0 {
		last = last[index+1:]
	}
	filename := strings.TrimSpace(last)
	if filename == "" {
		return "upload"
	}
	return filename
}
